Implement tensor.align_as(other), change tensor.align_to(names)

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25843 Implement tensor.align_as(other), change tensor.align_to(names)**
* #25842 Implement tensor.refine_names

`tensor.align_to(*names)` permutes the dimensions of `tensor` and adds
additional 1-sized dimensions such that the output tensor has dimensions
in the same order as `names`. All dimensions of `tensor` must be
present in `names`, in addition, this function requires that all dims of
`tensor` be named.

`tensor.align_as(other)` is equivalent to
`tensor.align_to(*other.names)`.

I'm planning on changing `torch.align_tensors(*tensors)` to align closer
to these semantics because there didn't seem to be a clear use case for the old
semantics that preserve unnamed dimensions. That will come in a future
change.

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25843

Differential Revision: [D17255549](https://our.internmc.facebook.com/intern/diff/D17255549)

zou3519 committed:
Implement tensor.align_as(other), change tensor.align_to(names)

`tensor.align_to(*names)` permutes the dimensions of `tensor` and adds
additional 1-sized dimensions such that the output tensor has dimensions
in the same order as `names`. All dimensions of `tensor` must be
present in `names`, in addition, this function requires that all dims of
`tensor` be named.

`tensor.align_as(other)` is equivalent to
`tensor.align_to(*other.names)`.

I'm planning on changing `torch.align_tensors(*tensors)` to align closer
to these semantics because there didn't seem to be a clear use case for the old
semantics that preserve unnamed dimensions. That will come in a future
change.

Test Plan:
- new tests [namedtensor ci]

nairbv commented on aten/src/ATen/native/NamedTensor.cpp:
```diff
@@ -142,12 +142,36 @@ static Tensor align(const Tensor& tensor, DimnameList names, bool is_aligning_tw
+    auto it = std::find(names.begin(), names.end(), dim);
+    TORCH_CHECK(it != names.end(),
+        "align_to: Cannot find dim ", dim, " from Tensor", names,
+        " in desired alignment ", names, ".");
```
do we usually end our error messages in a period? I notice one of these errors does and one doesn't.

zou3519 replied:
Some of our error messages end in a period, some of them don't. I can make both of these match but we shouldn't block this PR on that if it's the only problem

nairbv replied:
yeah not a blocker just a nit/question, would be nice to have a consistent convention for us to follow though.

nairbv commented on aten/src/ATen/native/NamedTensor.cpp:
```diff
@@ -142,12 +142,36 @@ static Tensor align(const Tensor& tensor, DimnameList names, bool is_aligning_tw
+    const auto& dim = tensor_names[idx];
+    TORCH_CHECK(dim.is_normal(),
+        "align_to: All input dims must be named. Found unnamed dim at index ",
+        dim, " of Tensor", tensor_names);
```
I think you probably want a space after "Tensor"?

zou3519 replied:
Nope, no space necessary. It should look like 

```
of Tensor['N', 'C', 'H', 'W']
```

for example. I'll try to toss this into a helper function in a followup PR because I've noticed that I've done this a lot in the codebase.

nairbv replied:
ah right, I think I've been confused by this before too but makes sense.

zou3519 replied:
Yeah, I'll toss it onto my clean-up list.

zou3519 committed:
Update on "Implement tensor.align_as(other), change tensor.align_to(names)"


`tensor.align_to(*names)` permutes the dimensions of `tensor` and adds
additional 1-sized dimensions such that the output tensor has dimensions
in the same order as `names`. All dimensions of `tensor` must be
present in `names`, in addition, this function requires that all dims of
`tensor` be named.

`tensor.align_as(other)` is equivalent to
`tensor.align_to(*other.names)`.

I'm planning on changing `torch.align_tensors(*tensors)` to align closer
to these semantics because there didn't seem to be a clear use case for the old
semantics that preserve unnamed dimensions. That will come in a future
change.

Test Plan:
- new tests [namedtensor ci]

Differential Revision: [D17255549](https://our.internmc.facebook.com/intern/diff/D17255549)

zou3519 committed:
Update on "Implement tensor.align_as(other), change tensor.align_to(names)"


`tensor.align_to(*names)` permutes the dimensions of `tensor` and adds
additional 1-sized dimensions such that the output tensor has dimensions
in the same order as `names`. All dimensions of `tensor` must be
present in `names`, in addition, this function requires that all dims of
`tensor` be named.

`tensor.align_as(other)` is equivalent to
`tensor.align_to(*other.names)`.

I'm planning on changing `torch.align_tensors(*tensors)` to align closer
to these semantics because there didn't seem to be a clear use case for the old
semantics that preserve unnamed dimensions. That will come in a future
change.

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25843

Differential Revision: [D17255549](https://our.internmc.facebook.com/intern/diff/D17255549)

zou3519 committed:
Update on "Implement tensor.align_as(other), change tensor.align_to(names)"


`tensor.align_to(*names)` permutes the dimensions of `tensor` and adds
additional 1-sized dimensions such that the output tensor has dimensions
in the same order as `names`. All dimensions of `tensor` must be
present in `names`, in addition, this function requires that all dims of
`tensor` be named.

`tensor.align_as(other)` is equivalent to
`tensor.align_to(*other.names)`.

I'm planning on changing `torch.align_tensors(*tensors)` to align closer
to these semantics because there didn't seem to be a clear use case for the old
semantics that preserve unnamed dimensions. That will come in a future
change.

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25843

Differential Revision: [D17255549](https://our.internmc.facebook.com/intern/diff/D17255549)

zou3519 committed:
Update on "Implement tensor.align_as(other), change tensor.align_to(names)"


`tensor.align_to(*names)` permutes the dimensions of `tensor` and adds
additional 1-sized dimensions such that the output tensor has dimensions
in the same order as `names`. All dimensions of `tensor` must be
present in `names`, in addition, this function requires that all dims of
`tensor` be named.

`tensor.align_as(other)` is equivalent to
`tensor.align_to(*other.names)`.

I'm planning on changing `torch.align_tensors(*tensors)` to align closer
to these semantics because there didn't seem to be a clear use case for the old
semantics that preserve unnamed dimensions. That will come in a future
change.

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25843

Differential Revision: [D17255549](https://our.internmc.facebook.com/intern/diff/D17255549)

zou3519 committed:
Update on "Implement tensor.align_as(other), change tensor.align_to(names)"


`tensor.align_to(*names)` permutes the dimensions of `tensor` and adds
additional 1-sized dimensions such that the output tensor has dimensions
in the same order as `names`. All dimensions of `tensor` must be
present in `names`, in addition, this function requires that all dims of
`tensor` be named.

`tensor.align_as(other)` is equivalent to
`tensor.align_to(*other.names)`.

I'm planning on changing `torch.align_tensors(*tensors)` to align closer
to these semantics because there didn't seem to be a clear use case for the old
semantics that preserve unnamed dimensions. That will come in a future
change.

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25843

Differential Revision: [D17255549](https://our.internmc.facebook.com/intern/diff/D17255549)
